app/controllers/imports_controller.rb

```diff
@@ -1,30 +1,74 @@
 class ImportsController < ApplicationController
+  include ActionView::Helpers::UrlHelper
+
   before_action :authenticate_user!
   before_action :must_be_admin
-  before_action :check_empty_attachment, only: [:create]
 
   def index
     @import_type = params.fetch(:import_type, "volunteer")
+    @import_error = session[:import_error]
+    session[:import_error] = nil
   end
 
   def create
     import = import_from_csv(params[:import_type], params[:file], current_user.casa_org_id)
-    flash[import[:type]] = import[:message]
+    message = import[:message]
+
+    # If there were failed imports
+    if import[:exported_rows]
+      message << "<p class='mt-4'>" + link_to("Click here to download failed rows.", download_failed_imports_path) + 
+        "</p>" + "<p>#{ERR_FAILED_IMPORT_NOTE}</p>"
+      session[:exported_rows] = import[:exported_rows]
+    end
+
+    if import[:type] == :error
+      session[:import_error] = message
+    # Only use flash for success messages. Otherwise may cause CookieOverflow
+    else
+      flash[:success] = message
+    end
+
     redirect_to imports_path(import_type: params[:import_type])
   end
 
+  def download_failed
+    data = session[:exported_rows]
+    session[:exported_rows] = nil
+    send_data data, format: :csv, filename: "failed_rows.csv"
+  end
+
   private
 
+  ERR_FAILED_IMPORT_NOTE = "Note: An additional 'error' column has been added to the file. " \
+    "Please note the failure reason and remove the column when resubmitting."
+  ERR_FILE_NOT_ATTACHED = "You must attach a CSV file in order to import information!"
+  ERR_FILE_NOT_FOUND = "CSV import file not found."
+  ERR_FILE_EMPTY = "File can not be empty."
+  ERR_INVALID_HEADER = "Looks like this CSV contains invalid formatting. " \
+    "Please download an example CSV for reference and try again."
+
+  def header
+    {
+      "volunteer" => ["display_name", "email"],
+      "supervisor" => ["email", "display_name", "supervisor_volunteers"],
+      "casa_case" => ["case_number", "transition_aged_youth", "case_assignment", "birth_month_year_youth"]
```

Comment:
We could probably wrap these arrays into constants within the individual classes. e.g. `VolunteerImporter::IMPORT_HEADERS` and then reuse those in the class itself (and tests).

Reply:
bab1e0f5a7ab9cfe4d0733799cf466fae32c972a
Also added a few specs for the header validation